src/mongo/client/dbclient.cpp

```diff
@@ -1066,15 +1092,153 @@ namespace mongo {
         return runCommand(db.c_str(), b.done(), *info);
     }
 
-    bool DBClientWithCommands::copyDatabase(const string &fromdb, const string &todb, const string &fromhost, BSONObj *info) {
+    bool DBClientWithCommands::copyDatabase(const string& fromdb,
+                                            const string& todb,
+                                            const string& fromhost,
+                                            const string& mechanism,
+                                            const string& username,
+                                            const string& password,
+                                            BSONObj *info) {
         BSONObj o;
         if ( info == 0 ) info = &o;
-        BSONObjBuilder b;
-        b.append("copydb", 1);
-        b.append("fromhost", fromhost);
-        b.append("fromdb", fromdb);
-        b.append("todb", todb);
-        return runCommand("admin", b.done(), *info);
+        BSONObjBuilder copydbCmd;
+        copydbCmd.append("copydb", 1);
+        copydbCmd.append("fromhost", fromhost);
+        copydbCmd.append("fromdb", fromdb);
+        copydbCmd.append("todb", todb);
+
+        // If we don't have a username, or if we're copying locally,
+        // just run the command without authenticating
+        if ( username == "" || fromhost == "" ) {
+            return runCommand("admin", copydbCmd.done(), *info);
+        }
+
+        // Otherwise, take or guess the auth mechanism
+        string authMech;
+        if ( mechanism.c_str() != kAuthMechDefault ) {
+            uassert( 0, "auth mechanism must be MONGODB-CR or SCRAM-SHA-1",
+                     ( mechanism.c_str() != kAuthMechMongoCR &&
+                       mechanism.c_str() != kAuthMechScramSha1));
```

Comment:
two things, first i think the logic here is inverted, second, as above you don't need c_str()

shouldn't this be
`uassert( 0, ..., (mechanism == kAuthMongoCR) || (mechanism == kAuthMechScramSha1)`)